rfcs/20250416-gemm-external-zp-precompute/README.md

```diff
@@ -0,0 +1,87 @@
+# User-Supplied Precomputed Zero Points for GEMM
+
+## Introduction
+
+For modern LLM networks, which usually are a bunch of relatively small GEMMs, performance is priority #1.
+To optimize LLMs for performance (i.e. use DPAS and DP4A, the hardware matmul instructions of Intel GPUs)
+a novel approach has been proposed which is to dynamically downcast both A and B matrices from a floating
+point type to INT8 — but to avoid losing too much precision the downcasting algorithm should also produce
+zero points.
+
+Of particular interest is the variation where the B matrix, or "weights", which is usually constant, gets
+downcast ahead of time. The problem is that zero points on B mean that A (the "sources") is to be reduced
+by the K dimension to multiply the resulting reductions to the ZPs extracted from B; see formulas below.
+
+Calculating partial K reductions of A within the GEMM kernel would never be efficient; there is, however,
+a way to perform them painlessly while downcasting A. Everything that needs adding to GEMM to use them is
```

Comment:
Note that oneDNN does not provide a reduction post-op, so the user will have to fuse downcast+reduction to the previous kernel themselves. Is that ok for the target usage?

Reply:
Yes, as the OpenVINO team who proposed the op already have this primitive implemented.

Reply:
To be precise, it's not a "primitive" but a custom kernel that performs dynamic quantization of the tensor A which needs to compute max for scales, I presume, and as it does so, will also collect a partial sums values into a dedicated tensor.